packages/@react-aria/datepicker/src/useDateSegment.ts

```diff
@@ -256,7 +256,8 @@ export function useDateSegment(segment: DateSegment, state: DateFieldState, ref:
           if (shouldSetValue) {
             focusManager.focusNext();
           }
-        } else {
+        } else if (shouldSetValue) {
+          // Don't accept leading zeros except for fields that accept 0 as a entire value (aka 00 for minutes/seconds/etc)
```

Comment:
If we wanna constraint this behavior to just the year field, can do something like 
```
} else if (!(segment.type === 'year' && !shouldSetValue)) {
``` 
instead. Would make it inconsistent with the other fields but could potentially be argued for since people don't really type leading zeros when entering years